[sjaqjwor(이승기)] step5 pr 요청드립니다.

sjaqjwor opened the pull request:
안녕하세요

wheejuni님 step5 pr 요청드립니다.

이전 pr때 실수로 push 안하고 PR 드린점 정말로 죄송합니다... 😢 

이번 pr에는 저번에 피드백을 받지 못하고 누락된 Commit CherryPick 하여 넣었습니다.

이와 덜불어 step5기능도 추가했습니다.

피드백 부탁드리겠습니다

정말 감사드립니다.

sjaqjwor committed:
refactor(prize) : Prize 스스로 개수를 검사하고 결과값을 반환하다록 변경

sjaqjwor committed:
refactor(prize) : prizeOf 내부 로직 깔끔하게 정리

sjaqjwor committed:
feat (prizes) : prize리스트를 가지고 있는 일급 컬렉션 및 상품 개수 반환 ,총 보상금 기능 추가

sjaqjwor committed:
refactor(Lottos) : getPrizeOfLotto 메소드 반환값 변경 및 로직 정리

sjaqjwor committed:
refactor (lottoBuyer) : convertStatisticResult 매개변 수 변경에 따른 로직 및 테스트 코드 수정

sjaqjwor committed:
refcator (UI) : getPrizeOfLotto 결과 리턴값 변경으로 인한 UI 변경

sjaqjwor committed:
refactor(Money) : Money에서 investment라는 불필요 지역변수 제거

sjaqjwor committed:
refactor(lottoSeller) : 가격 정보를 가지고 있으면서 외부로 공개

sjaqjwor committed:
refactor(LottoBuyer) : 수익률을 계산 할 때 외부에서 로또 개수를 받도록 변경

sjaqjwor committed:
refactor(winnerLotto) : winnerLotto의 상위 타입을 Lotto로 지정 및 Lotto에서 필요한 메소드 override

sjaqjwor committed:
refactor(Lotto) : Lotto Prize 반환하는 메소드 추가 및 하위 클래스에  LottoNumber 포함여부 메소드 개방

sjaqjwor committed:
refactor (LottoManger) : WinnerLotto객체가 아닌 Lotto객체로 할당

sjaqjwor committed:
docs (readme) : 요구 사항 분석 및 개발 기능 문서 정리

sjaqjwor committed:
refactor (LottoNumberGenerator) : 수동 로또 번호를 LottoNumber로 변환하여 반환하는 기능 제공

sjaqjwor committed:
refactor (lottoSeller) : 수동 로또 판매 기능 추가 및 구매 가능 수량 반환 메소드 추가

sjaqjwor committed:
refactor (LottoBuyer) : 수동 로또 번호를 매개변수로 받아 수동 로또와 자동 로또를 구매하는 기능 제공

sjaqjwor committed:
refactor(Lottos) :  자동 로또와 수동 로또를 매개변수로 받는 생성자 생성 후 하나로 합치는 기능 제공

sjaqjwor committed:
refactor(UI 및 Manager) : 수동 로또 view , output 제공

sjaqjwor committed:
refactor : 코드 indent 및 사용하지 않는 import 정리

wheejuni commented on src/main/java/lotto/domain/buyer/LottoBuyer.java:
```diff
@@ -4,32 +4,34 @@
+    public BigDecimal calculateYield(Prizes prizes, int lottoCount) {
+        BigDecimal rewardSum = prizes.sumReward();
+        return rewardSum.divide(BigDecimal.valueOf(lottoCount * LottoSeller.LOTTO_PRICE), DECIMAL_POINT_POSITION, BigDecimal.ROUND_FLOOR);
+    }
```
`Prizes` 에게 `lottoCount` 를 알려주면 수익률이 계산되는 고급 구현이 가능할 듯 합니다.

wheejuni commented on src/main/java/lotto/domain/lotto/Lottos.java:
```diff
@@ -15,14 +17,18 @@ public Lottos(List<Lotto> lottos) {
+        List<Prize> resultPrizes = new ArrayList<>();
         for (Lotto lotto : lottos) {
-            int count = lotto.matchCount(winnerLotto);
-            boolean hasBonusBall = lotto.hasBonusNumber(winnerLotto);
-            prizes.merge(Prize.prizeOf(count, hasBonusBall), 1, Integer::sum);
+            Prize prize = winnerLotto.matchPrize(lotto);
+            resultPrizes.add(prize);
```
stream API로 개선 시도해주세요.

sjaqjwor replied:
넵넵! 제가 stream에 익숙하지 않아서 for 문으로 했는데
stream으로 수정해보도록 하겠습니다.

한가지 궁금한 사항이 있습니다!!

궁금한 부분은 지금 로직이아니라

같은 로직이여도 for vs stream을

선택하는 기준이 있을까요??

sjaqjwor replied:
stream API로 수정하겠습니다

😄

wheejuni commented on src/main/java/lotto/prize/Prizes.java:
```diff
@@ -0,0 +1,37 @@
+
+public class Prizes {
+
+    private final List<Prize> prizes;
```
대부분의 경우에 `List<T>` 에 대해 `final` 을 붙인다고 불변성이 보장되지는 않습니다.
더 안전한 방법은 무엇이 있을지 고민해보시는 것도 좋습니다.

sjaqjwor replied:
넵 한번 고민하고 적용하겠습니다 😄

wheejuni commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -22,4 +24,19 @@ public static String inputBonusNumber() {
+    public static List<String> inputManualLottoNumbers(int manualQuantity) {
+        System.out.println("수동으로 구매할 번호를 입력해 주세요.");
+        List<String> manualNumbers = new ArrayList<>();
+        for (int qunatity = 0; qunatity < manualQuantity; qunatity++) {
```
오타가 있네요.

sjaqjwor replied:
qunaitity 모르는 단어네요...

quantity 로 수정하겠습니다

감사합니다!

wheejuni reviewed (changes requested):
간단한 제안 몇 가지 드렸습니다 👍 
확인 부탁드려요.

sjaqjwor committed:
refatcort  : 변수 이름 오타 수정 quantity 수정

sjaqjwor committed:
refactor(Prizes) : 로또 개수를 받아서 수익률을 계산하는 로직 추가 및 prizes 생성자 수정

update :

Prizes 생성자에서 prizes를 할당할때 Collections.unmodifiableList(new ArrayList<>(prizes)); 사용하여 원본 리스트가 변경해도
해당 prizes에는 변경이 없도록 수정

sjaqjwor committed:
refactor(LottoBuyer) : 단순 구매하는 기능만 남기고 수익률 계산 기능은 prizes로 이동

sjaqjwor committed:
refactor(lottos) : 생성자에서 로또 리스트 받아서 할당 시 변경 불가능한 리스트로 할당하도록 수정

sjaqjwor committed:
refactor(lottos) : 생성자에서 로또 넘버 리스트 받아서 할당 시 변경 불가능한 리스트로 할당하도록 수정

sjaqjwor committed:
refactor(LottoManager) : prizes에게 수익률 계산을 위임하면서 수익률 계산하는 부분 수정

sjaqjwor commented:
@wheejuni 님 안녕하세요

step5 피드백 주신거 반영해 보았습니다.

피드백 부탁드립니다.

## 수익률 계산
- LottoBuyer는 단순 로또를 구매하는 역할만 수행하도록 변경 (수익률 계산 제거)
- Prizes에게 수익률 계산 역할 부여

## final LIst<T> 에 할당하는 방식 개선
```
public Prizes(List<Prize> prizes) {
        ....
        this.prizes = Collections.unmodifiableList(new ArrayList<>(prizes));
    }
```
Collections.unmodifiableList로 한번더 매핑을 했습니다.
### new ArrayList<>(prizes)
한번도 감싼 이유는 원본으로 들어오는 prizes가 다른 로직에서 추가가 되면
 Collections.unmodifiableList만 감싼 리스트에도 변경이 오기 때문에 new ArrayList<>()로 한번더
매핑을 했습니다.

wheejuni reviewed (approved):
수고 많으셨습니다 👍 리뷰 사항이 많았는데 모두 잘 반영해 주셨습니다.
